aten/src/ATen/native/mkldnn/xpu/detail/WoQMatmul.cpp

```diff
@@ -0,0 +1,306 @@
+#include <c10/xpu/XPUFunctions.h>
+
+#include <ATen/native/mkldnn/xpu/detail/Attr.h>
+#include <ATen/native/mkldnn/xpu/detail/Utils.h>
+
+#include <oneapi/dnnl/dnnl.hpp>
+#include <cstdint>
+
+
+namespace at::native::onednn {
+
+inline Tensor resize_as_onednn_mat1(const Tensor& mat1, const Tensor& output) {
+  auto output_ = output.flatten(0, -2);
+  int n = output_.sizes()[1];
+  auto sizes = mat1.sym_sizes().vec();
+  sizes[sizes.size() - 1] = n;
+  return output.view_symint(sizes);
+}
+
+sycl::event woq_matmul_int4(
+    Tensor& result, // dst, [M, N]
+    const Tensor& mat1_, // src, [M, K]
+    const Tensor& mat2_, // quantized weight, [K/8, N]
+    const Tensor& scale, // [K/group_size, N]
+    const Tensor& zp, // [k/group_size, N/8]
+    int64_t group_size,
+    bool m2_trans,
+    Attr attr,
+    const c10::optional<Tensor>& g_idx,
+    const std::vector<sycl::event>& deps,
+    Tensor b_raw = at::Tensor()) {
+  Tensor mat1;
+  if (g_idx.has_value()) {
+    mat1 = mat1_.index_select(-1, g_idx.value()).flatten(0, -2);
+  } else {
+    mat1 = mat1_.flatten(0, -2);
+  }
+  auto mat2 = mat2_.flatten(0, -2);
+  int m = mat1.sizes()[0];
+  int n = mat2.sizes()[1];
+  int k = mat1.sizes()[1];
+  result = at::empty({m, n}, mat1_.options());
+  size_t dims = result.dim();
+  TORCH_CHECK(
+      dims == 2 || dims == 3,
+      "oneDNN matmul only works with 2D or 3D, got ",
+      dims);
+  TORCH_CHECK(
+      dims == mat1.dim() && dims == mat2.dim(),
+      "oneDNN input matrixes must have the same ranks");
+  TORCH_CHECK(result.defined(), "oneDNN matmul result should be defined");
+
+  at::Device curDevice = at::Device(at::kXPU, at::xpu::current_device());
+  auto engine = GpuEngineManager::Instance().get_engine(curDevice);
+  auto engine_index = curDevice.index();
+  auto stream = GpuStreamManager::Instance().get_stream();
+
+  // make them all contiguous
+  Tensor m1 = is_onednn_matmul_strides(mat1) ? mat1 : mat1.contiguous();
+  Tensor m2 = is_onednn_matmul_strides(mat2) ? mat2 : mat2.contiguous();
+  Tensor scale_ = is_onednn_matmul_strides(scale) ? scale : scale.contiguous();
+  Tensor zp_ = is_onednn_matmul_strides(zp) ? zp : zp.contiguous();
+  Tensor dst = is_onednn_matmul_strides(result, true) ? result : result.contiguous();
+
+  m = dst.size(-2);
+  n = dst.size(-1);
+  k = m1.size(-1);
+  int64_t mb = 1;
+
+  if (dims == 3) {
+    mb = dst.size(0);
+    TORCH_CHECK(
+        mb == m1.size(0) && mb == m2.size(0),
+        "batch size mismatch, dst mb: ",
```

Comment:
Thanks for your suggestions, `mb` means minibach here, but  i review the code and remove `mb` in the code since int4_gemm has no need to handle batch currently.